Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp

```diff
@@ -0,0 +1,257 @@
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "URLPatternCanonical.h"
+
+#include <wtf/text/MakeString.h>
+#include <wtf/text/StringToIntegerConversion.h>
+
+namespace WebCore {
+
+static bool isInvalidIPv6HostCodePoint(auto codepoint)
+{
+    static constexpr std::array<LChar, 3>  invalidSpecialCodepoints =  { '[', ']', ':' };
+    return !isASCIIHexDigit(codepoint) && std::find(invalidSpecialCodepoints.begin(), invalidSpecialCodepoints.end(), codepoint) != invalidSpecialCodepoints.end();
+}
+
+// https://urlpattern.spec.whatwg.org/#is-an-absolute-pathname
+bool isAbsolutePathname(StringView input, BaseURLStringType inputType)
+{
+    if (input.isEmpty())
+        return false;
+
+    if (input.startsWith('/'))
+        return true;
+
+    if (inputType == BaseURLStringType::URL)
+        return false;
+
+    if (input.length() < 2)
+        return false;
+
+    if (input.startsWith("\\/"_s))
+        return true;
+
+    if (input.startsWith("{/"_s))
+        return true;
+
+    return false;
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-a-protocol, combined with https://urlpattern.spec.whatwg.org/#process-protocol-for-init
+ExceptionOr<String> canonicalizeProtocol(const String& value, BaseURLStringType valueType)
+{
+    if (value.isEmpty())
+        return String { value };
+
+    auto strippedValue = value.endsWith(':') ? value.left(value.length() - 1) : value;
+
+    if (valueType == BaseURLStringType::Pattern)
+        return strippedValue;
+
+    URL dummyURL(makeString(strippedValue, "://webkit.test"_s));
+
+    if (!dummyURL.isValid())
+        return Exception { ExceptionCode::TypeError, "Invalid input to canonicalize a URL protocol string."_s };
+
+    return dummyURL.protocol().toString();
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-a-username, combined with https://urlpattern.spec.whatwg.org/#process-username-for-init
+String canonicalizeUsername(const String& value, BaseURLStringType valueType)
+{
+    if (value.isEmpty())
+        return String { value };
+
+    if (valueType == BaseURLStringType::Pattern)
+        return String { value };
+
+    URL dummyURL("https://www.webkit.org"_s);
+    dummyURL.setUser(value);
+
+    return dummyURL.user();
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-a-password, combined with https://urlpattern.spec.whatwg.org/#process-password-for-init
+String canonicalizePassword(const String& value, BaseURLStringType valueType)
+{
+    if (value.isEmpty())
+        return String { value };
+
+    if (valueType == BaseURLStringType::Pattern)
+        return String { value };
+
+    URL dummyURL("https://www.webkit.org"_s);
```

Comment:
`"https://www.webkit.org"_s` is used in several places. Ideally we'd have a global constant for it:
```
static constexpr auto dummyURLCharacters { "https://www.webkit.org"_s };
```